### ISSUE TITLE
PatternPaint crashes on ubuntu 15.04 when selecting color

### ISSUE THREAD

**Issue by rp10007** [Blinkinlabs/PatternPaint#123]
Using QT creator community edition. Configured for desktop Qt 5.5.1, GCC 64bit. I can amke the color picker appear, but as soon as I select something or try to change one of the values in the hsv or rgb boxes everything goes away. It's quite possible I've done something stupid (I don't know Qt at all) but no idea what.

Here's what the general messages window says:
Starting /home/paul/PatternPaint/build-PatternPaint-Desktop_Qt_5_5_1_GCC_64bit-Debug/PatternPaint...
debug MainWindow::connectionScannerTimer_timeout: BlinkyTapes found: 1
debug MainWindow::connectionScannerTimer_timeout: Attempting to connect to tape on: "ttyACM0"
debug BlinkyTape::open: Connecting to device on  "ttyACM0"
debug BlinkyTape::handleBaudRateChanged: Baud rate updated to 115200!
debug MainWindow::on_blinkyConnectionStatusChanged: status changed, connected= true
warning unknown: QPainter::begin: Paint device returned engine == 0, type: 3
warning unknown: QPainter::end: Painter not active, aborted
warning unknown: QPainter::begin: Paint device returned engine == 0, type: 3
warning unknown: QPainter::end: Painter not active, aborted
The program has unexpectedly finished.
/home/paul/PatternPaint/build-PatternPaint-Desktop_Qt_5_5_1_GCC_64bit-Debug/PatternPaint crashed

**Comment by cibomahto** [comment]
Interesting. I haven't tried Qt5.5.1 on Linux yet. Have you installed Qt from your distro's package or the official download from their site? And which flavor and version of OS are you running?

I'm on a low bandwidth connection at the moment but will download qt 5.5.1 tomorrow and try to reproduce this.

The 'warning unknown: QPainter:' messages are happening on all platforms and I think are a separate issue (I'll file that one now)

**Comment by rp10007** [comment]
Ubuntu 15.04. I installed from distro, got the Qtserial problem, then installed the official download. The official download version worked (except for the warning and the crash). So I tried to get it to work with similar settings on the distro version (5.4.1 I think) but managed to hit something wrong on the configuration menu so that the distro version is completely borked (until I figure out how to tell it where the desktop kit and cmake and a bunch of other things are).

So tl;dr: official download version.

**Comment by cibomahto** [comment]
Just tried Qt 5.5.1 on my 14.04 box, and wasn't able to reproduce the issue, so I would guess it is a configuration issue.
![screen shot 2015-12-14 at 1 05 25 am](https://cloud.githubusercontent.com/assets/44493/11768147/ea710706-a1fe-11e5-9140-2d9f7fcb3529.png)

Can you try deleting the PatternPaint repo that you downloaded, download it again, and open it in Qt Creator?

Alternatively, try deleting the 'build-PatternPaint-Desktop-Debug' and any other build-PatternPaint-* folders inside of the repo directory (all the temporary build files and executables are stored there), as well as any PatternPaint.pro.user* files in the PatternPaint/ subdirectory. That should effectively wipe any build kit selections that were stored in the repo.

If the problem persists, try running it in debug mode (debug->start debugging->start debugging) and recreate the issue. Qt Creator should automatically switch to a debug view that shows the stack trace in the bottom middle- right click on that and it should have an option to copy it to clipboard, then paste it here.
![thread stack](https://cloud.githubusercontent.com/assets/44493/11768206/5bab33e6-a200-11e5-859a-973cdd97fef1.png)

**Comment by rp10007** [comment]
Finally got back to this. I an in debug mode and got a segv with the following stack trace:

0	_int_malloc	/usr/lib/debug/lib/x86_64-linux-gnu/libc-2.21.so	3326	0x7ffff565a98d	
1	_int_realloc	/usr/lib/debug/lib/x86_64-linux-gnu/libc-2.21.so	4271	0x7ffff565c961	
2	__GI___libc_realloc	/usr/lib/debug/lib/x86_64-linux-gnu/libc-2.21.so	3023	0x7ffff565db0b	
3	g_realloc			0x7ffff488c85e	
4	??			0x7fffe7d7b6fd	
5	??			0x7fffe7d7c60a	
6	gdk_region_union			0x7fffe7d7ccc5	
7	??			0x7fffe7d876e0	
8	??			0x7fffe7d87856	
9	gtk_widget_queue_draw_area			0x7fffec24bdba	
10	gtk_widget_queue_draw			0x7fffec252d23	
11	??			0x7fffec0a4794	
12	??			0x7fffec0a47de	
13	??			0x7fffec65c434	
14	ColorChooser::setColor	colorchooser.cpp	68	0x431853	
15	ColorChooser::setAndSendColor	colorchooser.cpp	79	0x4318c0	
16	ColorChooser::qt_static_metacall	moc_colorchooser.cpp	83	0x455aca	
17	QMetaObject::activate(QObject *, int, int, void * *)			0x7ffff649035e	
18	QColorDialog::currentColorChanged(QColor const&)			0x7ffff767ff92	
19	??			0x7ffff7686379	

(This is when clicking on the color square and then clicking in the color picker to change one of the rgb values; if I click on the eyedropper and try to pick a color that way the crash leaves the cursor in a state that won't click to execute any commands, so I can't easily get the stack trace.)

hope this helps diagnose things

**Comment by cibomahto** [comment]
Hmm, looking at the last commit from a few days ago:
https://github.com/Blinkinlabs/PatternPaint/blob/0eaacbbbea7fac580e629a374197bc97df357aa2/PatternPaint/colorchooser.cpp

line 68 is: setPixmap(pixmap);

which is an internal function to QLabel:
http://doc.qt.io/qt-5/qlabel.html#pixmap-prop

Assuming that is the fault, it's frustrating but not a showstopper. I made an alternate implementation of the ColorChooser widget based on a QWidget, and sidestepped the pixmap stuff by just implementing a custom draw function. Try this branch:
https://github.com/Blinkinlabs/PatternPaint/tree/alternateColorChooser

and see if it helps... Details here:
https://github.com/Blinkinlabs/PatternPaint/commit/eaaadffa92dd4cc653026aeec15a8e6ffc4b4464

**Comment by rp10007** [comment]
Alas no: stack trace on segfault below. Oh, and another minor bit: at some point a typo seems to have gotten into blinkycontroller.cpp: QPOinter instead of QPointer.  Thanks for the ongoing work!

1	___printf_fp	printf_fp.c	501	0x7ffff5685c7e	
2	_IO_vfprintf_internal	vfprintf.c	1660	0x7ffff5684793	
3	___vsnprintf_chk	vsnprintf_chk.c	63	0x7ffff573f235	
4	g_snprintf			0x7ffff4d380d2	
5	??			0x7fffed616727	
6	??			0x7fffedbbafe0	
7	ColorChooser::setColor	colorchooser.cpp	75	0x42ea99	
8	ColorChooser::setAndSendColor	colorchooser.cpp	85	0x42eae0	
9	ColorChooser::qt_static_metacall	moc_colorchooser.cpp	82	0x4541cc	
10	QMetaObject::activate(QObject *, int, int, void * *)			0x7ffff6524ace	
11	QColorDialog::currentColorChanged(QColor const&)			0x7ffff768e8b2	
12	QMetaObject::activate(QObject *, int, int, void * *)			0x7ffff6524ace	
13	QPlatformColorDialogHelper::currentColorChanged(QColor const&)			0x7ffff6fe9ff2	
14	??			0x7fffedbba7e2	
15	??			0x7ffff4fc655f	
16	g_signal_emit_valist			0x7ffff4fdf088	
17	g_signal_emit			0x7ffff4fdfce2	
18	??			0x7fffed61686a	
19	??			0x7fffedbbafe0	
20	ColorChooser::setColor	colorchooser.cpp	75	0x42ea99	
...	<More>				

Also: I'm getting a warning:
/home/paul/PatternPaint-alternateColorChooser/PatternPaint/main.cpp:38: warning: #warning Updater initialization disabled [-Wcpp]
 #warning Updater initialization disabled
  ^

**Comment by cibomahto** [comment]
The warning is safe to ignore- it's a reminder that the mechanism to check for updates is disabled in debug mode. Technically it's useless since the updater isn't implemented on Linux.

**Comment by cibomahto** [comment]
It looks like it is getting stuck in an endless loop passing a color changed notification between the system color dialog and the color widget. Can you post a longer version of the stack trace? I'll try making a version of the widget with separate entry points for the notifications to prevent this kind of looping.

Perhaps something changed in a newer version of the GTK widget, or the Qt wrapper around it, that prevents it from filter events as on other platforms.

**Comment by cibomahto** [comment]
Ok, try this one:
https://github.com/Blinkinlabs/PatternPaint/tree/alternateColorChooser

Details here:
https://github.com/Blinkinlabs/PatternPaint/commit/6c1ec8d726e254717c330d2fb49e328448d170d5

**Comment by rp10007** [comment]
DingDingDing! Or at least I've tested it one one machine where the previous PatternPaints were crashing reliably and this one doesn't crash with repeated use of the colorpicker dialog and of the eyedropper. Sometime this weekend I'll try it on a couple other machines, but this is looking good so far. Hurrah! and thanks!

**Comment by cibomahto** [comment]
Excellent! I'll test it on Windows this afternoon, and pull it in to main if there aren't any issues. I'll file this as a latent bug in PatternPaint that is exposed by an API change in an external library ;-)

**Comment by cibomahto** [comment]
Looks fine on Windows so I pulled it in. Please re-open this or create a new issue if you find another system that exhibits the same behavior
